woocommerce/woocommerce-gutenberg-products-block
# Prevent third party blocks from breaking the entire checkout/cart block

If a third party block registers itself on the checkout/cart block (or one of its' inner blocks) it will cause the entire Checkout/Cart block to error. See #5101 for more details.

This PR ads an empty `<BlockErrorBoundary>` component around each of the inner blocks rendered within the checkout/cart components. This will prevent the entire checkout/cart block from erroring if one of the sub components throws and error.

There is a risk here that because the error boundary doesn't show any errors and fails silently, that it may cause some difficulty in debugging. I think from a user perspective this is the best solution, as we don't want to show any cryptic error messages at the critical stage of a customer's journey. From a dev perspective, the error is thrown in the console and If a whole block goes missing from your checkout, you will know something is wrong! Would love to hear other's thoughts on this.

Further context in this slack p1638436874157200-slack-C8X6Q7XQU


Fixes #5101 

### Testing

### Manual Testing

How to test the changes in this Pull Request:

1. You will need two plugins that add a checkbox within the Contact Information inner block on the checkout. You can Install the [Newsletter test](https://github.com/woocommerce/newsletter-test) plugin and either duplicate it or install Mailpoet.
2. Visit the Checkout page. Expect to see two checkboxes under the email input field added by the newsletter-test and mailpoet plugins. They are rendered here as forced blocks, as the HTML hasn't been saved to the checkout yet.
3. Throw an error anywhere in this [Block](https://github.com/woocommerce/newsletter-test/blob/main/assets/js/checkout-newsletter-subscription-block/block.js#L10)
4. Reload the Chekout page. 
5. The newsletter checkbox should not be visible and the rest of the checkout block should display and function as normal. Specifically, note that the sibling checkbox is not affected. If you are logged in as an admin, you should see an error message in place of the erroring checkbox.
6. Now go to the Block based Checkout page. Make a small change here and click "Update".
7. Repeat steps 2-5. This time, the markup of the two checkbox blocks will be saved in the database, and the blocks will not be "forced" anymore.

### Edit: Additional Bug fix
While reviewing the PR, @opr pointed out a bug where the label of the checkbox from the `newsletter-subscription` plugin would disappear. Turns out this was to do with text nodes not being rendered properly within the `renderInnerBlocks()` function, which was expecting children only of type `Element`. I've changed the `renderInnerBlocks()` and `renderForcedBlocks()` functions to work with any `Node` rather than `Element`, and added the necessary types and type guards to prevent text nodes from accessing any properties only available on elements. I learned the difference between nodes and elements doing this which I wasn't aware of:

> An element is a specific type of node, one that can be directly specified in the HTML with an HTML tag and can have properties like an id or a class. can have children, etc. Nodes vs Elements: Nodes are all the different components that a webpage is made up of and elements are one type of node 

### Changelog

> Prevent third party blocks from breaking the entire checkout/cart blocks
